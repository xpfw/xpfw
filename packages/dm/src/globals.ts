const StatType = {
  sum: "sum",
  mean: "mean",
  timeStep: "timeStep",
  sumTimeDistance: "sumTimeDistance",
  sumPrevTimeDistance: "sumPrevTimeDistance",
  avgTimeDistance: "avgTimeDistance",
  avgPrevTimeDistance: "avgPrevTimeDistance"
}

export {
  StatType
}
