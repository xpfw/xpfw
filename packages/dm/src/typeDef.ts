export interface IStatConfig {
  type: string
  /**
   * used for recognition in react elements
   */
  id: string
  options?: any
}

export type IFindMethod = any

/**
 * Based StatCalculator Type
 * O is the type of the iterated Object
 * R is the type of the result of the calculator
 */
export type TypedStatCalculator<O, R> =
  (object: O) => Promise<R>

/**
 * Common Options for calculators
 */
export interface ICommonOptions {
  itemPath?: string
}

/**
 * Common Options for comparers
 * Will use itemPath and compareWithPath
 */
export interface ICompareOptions extends ICommonOptions {
  /**
   * path for compare with
   */
  compareWithPath?: string
}

/**
 * Common Options for timestep functions
 */
export interface ITimestepOptions extends ICommonOptions {
  /**
   * path for date to check
   */
  datePath?: string
  subConfig?: any
  subType?: string
  timeSteps: any
}

export interface IChartConfig {
  id: string
  calcType: string
  displayType: string
  options?: any
}

export type NumberCalculator = (config?: ICommonOptions,
                                findMethod?: (c: string, q: any) => Promise<any>) => TypedStatCalculator<any, number>

export type CompareNumberCalculator = (config?: ICompareOptions,
                                       findMethod?: (c: string, q: any) => Promise<any>)
                                       => TypedStatCalculator<any, number>

export type TimeStepCalculator = (config?: ITimestepOptions,
                                  findMethod?: (c: string, q: any) => Promise<any>)
                                  => TypedStatCalculator<any, number>

export type DefaultStatCalculator = TypedStatCalculator<any, void>
