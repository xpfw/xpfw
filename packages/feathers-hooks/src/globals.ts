const Method = {
  Create: "create",
  Update: "update",
  Patch: "patch",
  Find: "find",
  Remove: "remove",
  Get: "get"
}

export {
  Method
}
