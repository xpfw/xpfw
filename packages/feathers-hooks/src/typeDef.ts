import { Hook } from "@feathersjs/feathers"
import { ExtendedJSONSchema } from "@xpfw/form"
import { IParameters, IPermissionSchema } from "@xpfw/permission"

export type ValidateHook = (form: ExtendedJSONSchema, custom?: IParameters) => Promise<Hook>
export type GeneralValidateHook = (form: ExtendedJSONSchema, method: string, custom?: IParameters) => Promise<Hook>

export type PermissionHook = (form: IPermissionSchema, custom?: IParameters) => Hook
export type GeneralPermissionHook = (form: IPermissionSchema, method: string, custom?: IParameters) => Hook
