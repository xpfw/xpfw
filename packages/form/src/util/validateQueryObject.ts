import {
  get, intersection,
  isNumber, keys, set,
  union } from "lodash"
import { ExtendedJSONSchema } from "../jsonschema"
import jsonValidator from "../jsonValidator"

const sameTypeOperators = ["$eq", "$gt", "$gte", "$lt", "$lte", "$ne"]
const arrayQueryOperators = ["$in", "$nin"]
const arrayWithSubFieldsOperators = ["$or", "$and"]
// TODO: all categories below logical query operators
const possibleQueryOperators = union(sameTypeOperators, arrayQueryOperators, arrayWithSubFieldsOperators)

// TODO:_ rewrite so not field but form is given
const validateQueryObject: (value: any, schema: ExtendedJSONSchema) => any = async (value, schema) => {
  const newObj: any = {}
  let hadFindKeys = false
  if (schema.properties != null) {
    for (const subSchemaKey of Object.keys(schema.properties)) {
      const subSchema: any = schema.properties[subSchemaKey]
      const fieldVal = get(value, String(subSchema.title))
      const sameTypeKeys = intersection(keys(fieldVal), sameTypeOperators)
      if (sameTypeKeys.length > 0) {
        hadFindKeys = true
        const fieldObj: any = {}
        for (const key of sameTypeKeys) {
          jsonValidator.validate(subSchema, fieldVal[key])
          if (jsonValidator.errors != null && jsonValidator.errors.length > 0) {
            throw new Error(JSON.stringify(jsonValidator.errors))
          }
          fieldObj[key] = fieldVal[key]
        }
        set(newObj, String(subSchema.title), fieldObj)
      }
      const arrayQueryKeys = intersection(keys(fieldVal), arrayQueryOperators)
      if (arrayQueryKeys.length > 0) {
        hadFindKeys = true
        const fieldObj: any = {}
        const arrayField = {type: "array", items: {type: subSchema.type}}
        for (const key of arrayQueryKeys) {
          jsonValidator.validate(arrayField, fieldVal[key])
          if (jsonValidator.errors != null && jsonValidator.errors.length > 0) {
            throw new Error(JSON.stringify(jsonValidator.errors))
          }
          fieldObj[key] = fieldVal[key]
        }
        set(newObj, String(subSchema.title), fieldObj)
      }
    }
  }

  const arraySubFieldOp = intersection(Object.keys(value), arrayWithSubFieldsOperators)
  if (arraySubFieldOp.length > 0) {
    hadFindKeys = true
    for (const key of arraySubFieldOp) {
      const arr = value[key]
      if (Array.isArray(arr)) {
        const newArr = []
        for (const v of arr) {
          newArr.push(await validateQueryObject(v, schema))
        }
        newObj[key] = newArr
      } else {
        throw new Error(`${key} should be an array but isn't`)
      }
    }
  }
  if (hadFindKeys === false) {
    jsonValidator.validate(schema, value)
    if (jsonValidator.errors != null) {
      const errors = jsonValidator.errors.filter((err) => err.keyword !== "required")
      if (errors.length > 0) {
        throw new Error(JSON.stringify(jsonValidator.errors))
      }
    }
    return value
  }
  if (isNumber(get(value, "$limit"))) {
    set(newObj, "$limit", get(value, "$limit"))
  }
  if (isNumber(get(value, "$skip"))) {
    set(newObj, "$skip", get(value, "$skip"))
  }
  return newObj
}

export { possibleQueryOperators, arrayWithSubFieldsOperators }
export default validateQueryObject
