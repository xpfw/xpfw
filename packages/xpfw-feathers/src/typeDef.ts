import { Hook } from "@feathersjs/feathers"
import { globals, IForm, IParameters } from "@xpfw/validate"

export type ValidateHook = (form: IForm, custom?: IParameters) => Hook
export type GeneralValidateHook = (form: IForm, method: string, custom?: IParameters) => Hook
