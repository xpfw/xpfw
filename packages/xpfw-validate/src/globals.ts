const RequiredType = {
  Always: 2,
  AllNotSet: 3,
  SetToVal: 4
}

const ErrorType = {
  invalidType: 3,
  required: 4,
  tooLow: 5,
  tooHigh: 420,
  notEqual: 6,
  emptyUpdate: 7
}

const FieldType = {
  Text: 2,
  Number: 3,
  Boolean: 4,
  Date: 5,
  Select: 6,
  Password: 7,
  Array: 10,
  Object: 12,
  RelationshipSingle: 13,
  RelationshipMulti: 14,
  Slider: 15,
  Location: 16
}

const Permission = {
    Guest: "guest",
    User: "user",
    Owner: "owner",
    Admin: "admin",
    Server: "server"
}

const Method = {
  Create: "create",
  Update: "update",
  Patch: "patch",
  Find: "find",
  Remove: "remove",
  Get: "get"
}

const StatType = {
  sum: "sum",
  mean: "mean",
  timeStep: "timeStep",
  sumTimeDistance: "sumTimeDistance",
  sumPrevTimeDistance: "sumPrevTimeDistance",
  avgTimeDistance: "avgTimeDistance",
  avgPrevTimeDistance: "avgPrevTimeDistance"
}

const DateType = {
  dateTime: 2,
  date: 3,
  time: 4
}

const options = {
  userCollection: "users",
  userIdPath: "_id"
}

export {
  StatType, RequiredType, ErrorType, FieldType, Permission, Method, DateType, options
}
export default {
  StatType, RequiredType, ErrorType, FieldType, Permission, Method, DateType, options
}
