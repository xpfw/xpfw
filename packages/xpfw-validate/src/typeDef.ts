export interface ISelectItem {
  label: string
  value: any
}

export type ValidateFunction = (value: any, options: IField, additionalParams: IParametersBase) => Promise<any>

export interface IValidateRequiredOptions {
  type: number
  model?: string | string[]
  values?: any
}

export type FieldConverter = (value: any) => any
export type FieldConvertFunc =
FieldConverter | {create?: FieldConverter, update?: FieldConverter, find?: FieldConverter}

export interface IValidateOptions {
  /**
   * Wether the field should be required and under which circumstances
   */
  required?: IValidateRequiredOptions

  /**
   * Minimum Value (for Number) or Length (for string, array, etc.)
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  min?: number

  /**
   * Maximum Value (for Number) or Length (for string, array, etc.)
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  max?: number

  /**
   * Step Size for number fields
   *
   * @type {number}
   * @memberof IValidateOptions
   */
  step?: number
  /**
   * Recursive ValidationOptions allow for defining Objects or Arrays in one Definition
   *
   * @type {IValidateOptions}
   * @memberof IValidateOptions
   */
  validate?: IValidateOptions
  customOpts?: any
  /**
   * Will check if current value equals the given <Field.mapTo> Key
   * Will check IParametersBase.formData at parameters.formData.<Field.mapTo>
   *
   * @type {string[]}
   * @memberof IValidateOptions
   */
  match?: string[]
  /**
   * Enhance your Validation by adding a custom Check Function
   *
   * @type {ValidateFunction}
   * @memberof IValidateOptions
   */
  custom?: ValidateFunction
  /**
   * For Arrays the expected Type of the content of the array
   * For Date it can be one of global.DateType
   * @type {(number | string)}
   * @memberof IValidateOptions
   */
  type?: number | string
  objectDef?: IField[]

  /**
   * Validate the current user has at least this perm level
   * in order to make ALL actions if number or a specific action
   */
  permission?: IPermTypes

  /** If value isn't set then set it to this one */
  defaultValue?: any

  /**
   * Use this to hide fields in specific contexts
   * Hidden of false / 0, shown if true
   */
  hide?: IFieldVisibilityDef
  /**
   * Customize what FormStore.getFormData produces on a per field basis
   * Like hide and permission this can be set globally or per Method
   * @type {FieldConvetFunc}
   * @memberof IValidateOptions
   */
  convert?: FieldConvertFunc
  /**
   * Collection to search in if Field has type RelationshipSingle or RelationshipMany
   */
  relationshipCollection?: string
  /**
   * Path to get name from if has  type RelationshipSingle or RelationshipMany
   */
  relationshipNamePath?: string
  /**
   * Path to get id from if has  type RelationshipSingle or RelationshipMany
   */
  relationshipIdPath?: string
}

export type SelectOptions = ISelectItem[]

export type SelectOptionGetter = (state: any) => SelectOptions

export type SelectOptionsField = SelectOptions | SelectOptionGetter

export interface IFieldVisibilityDef {
  create?: boolean
  update?: boolean
  find?: boolean
  remove?: boolean
}
export interface IField {
  type: string | number
  mapTo: string
  theme?: string
  validate?: IValidateOptions
}

export interface IFormSection {
  fields: IField[]
  name?: string
}

export interface IStatConfig {
  type: string
  /**
   * used for recognition in react elements
   */
  id: string
  options?: any
}

export interface IForm {
  model: string
  sections: IFormSection[]
  collection?: string
  icon?: any
  multiCollection?: string[]
  permissions?: IPermissionsForForm
  /**
   * Configure numerical xpfw-dm using stats
   * See `globals.StatType` for available types
   */
  stats?: IStatConfig[]
  options?: {
    /**
     * will add belongsTo in ui-shared so Permission.Owner can properly function
     */
    addBelongsTo?: boolean
    /**
     * add and display a createdAt date field
     */
    addCreatedAt?: boolean
    /**
     * path to id for various things
     */
    idPath?: string
    /**
     * Function that will return the custom built query
     *
     * @type {Function}
     */
    queryBuilder?: (form: IForm, prefix?: string) => any
    /**
     * The default sort for this collection
     */
    defaultSort?: any
    /**
     * Wether in the related search a $nin query on the idPath field should be included
     * Be sure to convert the strings into ObjectIds on the server side in case you are using MongoDb!
     */
    filterOutById?: boolean
  }
}

export interface IFormInstance extends IForm {
  getFields: IField[]
}

export interface IFieldSelect extends IField {
  selectOptions?: SelectOptionsField
}

export interface IFieldWithOptions<T> extends IField {
  additionalOptions?: T
}

/**
 * The Registry is needed to enable things like a subpage /select/<formModel>.<fieldMapTo>
 * It can then only with the knowledge of the string find out if this field exists
 * and then display or do whatever with it
 *
 * @export
 * @interface IRegistry
 */
export interface IRegistry {
  forms: {[index: string]: IForm}
  fields: {[index: string]: IField}
  validators: {[index: string]: ValidateFunction}
  registerField: (field: IField) => void
  findField: (mapTo: string) => IField | null
  registerForm: (form: IForm) => void
  findForm: (model: string) => IForm
  registerValidator: (name: string, validator: ValidateFunction) => void
  findValidator: (name: string) => ValidateFunction | null
}

/**
 * Parameters are supposed to be a universal way to pass relevant JS-refs
 * to write validation / permission / whatever functions that can try searching
 * offline data first / throw error if not connected / fetch data if missing via custom
 */
export interface IParametersBase {
  /**
   * The Field that you want to check
   *
   * @type {IField}
   * @memberof IParametersBase
   */
  field?: IField
  /**
   * The Form containing your Field
   * This is required for validatePermission checks
   */
  form?: IForm
  /**
   * Type of method so you can allow fields only in certain situations (find but not update etc.)
   *
   * @type {string}
   * @memberof IParametersBase
   */
  method?: string
  /**
   * If you are searching for a Document save it here
   *
   * @type {string}
   * @memberof IParametersBase
   */
  docId?: string
  doc?: any
  /**
   * Helper Variable you can set to indicate wether server requests are currently possible
   *
   * @type {boolean}
   * @memberof IParametersBase
   */
  isOffline?: boolean
  /**
   * Here you can pass any Object that will help you to achieve your task offline
   *
   * @type {*}
   * @memberof IParametersBase
   */
  offlineData?: any
  /**
   * This should contain the current state mapping of your form
   * e.g. {
   *  "LoginForm.username": "im user"
   * }
   * where LoginForm = IForm.model and username = IField.mapTo
   * @type {*}
   * @memberof IParametersBase
   */
  formData?: any
  /**
   * Form Registry to use for validation.
   *
   * @type {IRegistry}
   * @memberof IParametersBase
   */
  registry: IRegistry
  /**
   * validatePermission will check the ID of a User and search for it here via lodash.get
   */
  idPath?: string
  /**
   * Put the currently authenticated user in here for Permission Checks
   */
  currentUser?: any
  /**
   * validatePermission will check if the current userID equals this key of the object
   * given object {belongsTo: "myUserId"} and currentUser has "myUserId" as id this will return Permission.Owner
   */
  directOwnershipPath?: string
  /**
   * The Permissions Hook will try to feathers.get the object via this serviceName
   * The ID this is called with is gotten from hook at `docIdPath`
   */
  serviceName?: string
  /**
   * See `servicename`
   */
  docIdPath?: string
}

/**
 * Utiltily Interface so you can have type checked sub functions for your custom parameters
 */
export interface IParametersCustom<T> extends IParametersBase {
  custom?: T
}

/**
 * For when you don't want typechecked custom parameters :)
 */
export interface IParameters extends IParametersBase {
  custom?: any
}

export type getPermitted = (currentType: number, requiredType?: number) => Promise<boolean>

export type validate = (field: IField) => Promise<any>

export interface IErrorBase {
  type: number | string
}
export interface IError extends IErrorBase {
  custom?: any
}

export interface IErrorCustom<T> extends IErrorBase {
  custom?: T
}

export interface IFieldError {
  mapTo: string
  errors: IError[]
}
export interface IDiff {
  mapTo: string
  prev: any
  new: any
}

export interface IFormError {
    model: string
    errors: IFieldError[]
}

export type GetPermissions =
(currentUser: any, options: IParameters) => Promise<boolean | null>

export type GetPermissionCustom<T> =
(currentUser: any, options: IParametersCustom<T>) => Promise<boolean | null>

export interface IPermTypes {
  create?: string
  update?: string
  find?: string
  get?: string
  remove?: string
}

export interface IPermissionsForForm {
  /**
   * Specify a PermissionType as string
   */
  required?: IPermTypes
  /**
   * This will let you override the check by returning a boolean if the user should be allowed.
   * Return null if you wish the xpfw-native check based on `required` permissions to return its own result
   * That way you can limit your `check`-function to e.g. only return on create functions.
   */
  check?: GetPermissions | GetPermissionCustom<any>
  /**
   * Make sure that the result of a `validateForm` call has the current userId set at this key
   *
   */
  requireOwner?: string | IPermTypes
  /**
   * For certain types or custom checks you can pass options here
   * e.g. for the Role-check you can specify which roles need to be presented
   */
  custom?: any
  /**
   * If your users ID is Permissions.Server or Permissions.Admin then allow any action
   * Set to true to disable
   */
  disableAdminOverride?: boolean
}

export interface IUiClient {
  client: any
  connectTo: (url: string, options?: {
    authOptions?: any,
    makeAuth?: boolean,
    useRest?: boolean,
    userStore?: any
    /**
     * Pass the dbStore to get realtime updates
     */
    dbStore?: any
    /**
     * Realtime Updates will only be fetched for collections defined here
     */
    collections?: string[]
  }) => void
  disconnect: () => void,
  login: (loginData: any) => Promise<{user: any, custom?: any}>,
  register: (registerData: any) => Promise<any>,
  logout: () => Promise<any>,
  get: (collection: string, id: any) => Promise<any>,
  remove: (collection: string, id: any) => Promise<any>,
  create: (collection: string, createData: any) => Promise<any>,
  find: (collection: string, createData: any) => Promise<any>,
  patch: (collection: string, id: any, createData: any) => Promise<any>,
}
