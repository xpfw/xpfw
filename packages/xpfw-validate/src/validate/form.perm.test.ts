import { set } from "lodash"
import { FieldType, Method, Permission, RequiredType } from "../globals"
import ValidationRegistry from "../registry"
import {
  BooleanFieldWithMultiPermissions, FormWithPermissionsOnFields,
  NameFieldPublic, ObjectWithDifferingPermsPerKey
} from "../testUtil/defs"
import { IField, IForm, IParameters } from "../typeDef"
import validateForm from "./form"
import validateQueryObject from "./queryObject"

const form: IForm = FormWithPermissionsOnFields

const untypedValidateForm: any = validateForm

const params: IParameters = {registry: ValidationRegistry, method: Method.Find}

test("Verify validateForm respects Permission Settings", async () => {
  await expect(
    untypedValidateForm({guestName: "asdf", permbool: true}, form, {
      method: Method.Create
    })
  ).resolves.toMatchSnapshot("Strip permbool because not permitted")
  await expect(
    untypedValidateForm({guestName: "asdf", permbool: true}, form, {
      registry: ValidationRegistry,
      currentUser: {_id: Permission.Admin},
      method: Method.Create
    })
  ).resolves.toMatchSnapshot("Don't strip permbool because admin is permitted")
  await expect(
    untypedValidateForm({guestName: "asdf", permbool: true}, form, {
      registry: ValidationRegistry,
      currentUser: {_id: Permission.User},
      method: Method.Update
    })
  ).resolves.toMatchSnapshot("strip permbool because for update only server is permitted")
})
