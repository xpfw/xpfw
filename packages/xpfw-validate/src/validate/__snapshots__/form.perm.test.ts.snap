// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify validateForm respects Permission Settings: Don't strip permbool because admin is permitted 1`] = `
Object {
  "guestName": "asdf",
  "permbool": true,
}
`;

exports[`Verify validateForm respects Permission Settings: Strip permbool because not permitted 1`] = `
Object {
  "guestName": "asdf",
}
`;

exports[`Verify validateForm respects Permission Settings: strip permbool because for update only server is permitted 1`] = `
Object {
  "guestName": "asdf",
}
`;
